File: src/events/workflow_run.rs
use crate::errors::Error;
use crate::events::base::{Repository, WebhookMessage, Workflow, WorkflowRun};
use crate::events::Event;
use serde::Deserialize;

const IGNORED_WORKFLOWS: [&str; 3] = [
    "CodeQL",
    "Automatic Dependency Submission",
    "Dependabot Updates",
];

#[derive(Deserialize)]
pub struct WorkflowRunEvent {
    pub action: String,
    pub workflow: Workflow,
    pub workflow_run: WorkflowRun,
    pub repository: Repository,
}

impl Event for WorkflowRunEvent {
    fn handle(&self) -> Result<Option<WebhookMessage>, Error> {
        if self.action != "completed" {
            return Ok(None);
        }

        if let Some(conclusion) = &self.workflow_run.conclusion {
            if IGNORED_WORKFLOWS.contains(&self.workflow.name.as_str()) && conclusion == "success" {
                return Ok(None);
            }

            let emoji = if conclusion == "failure" {
                "<:catscream:1325122976575655936>"
            } else {
                "<:pepethinking:1330806911141941249>"
            };

            let branch_name = self
                .workflow_run
                .head_branch
                .as_deref()
                .unwrap_or("unknown");
            let formatted = format!(
                "{emoji} Workflow [{conclusion}](<{}>) on [{}](<{}>)/[{branch_name}](<{}/tree/{branch_name}>)",
                self.workflow_run.html_url,
                self.repository.name,
                self.repository.html_url,
                self.repository.html_url,
            );

            return Ok(Some(WebhookMessage {
                content: formatted,
                username: self.workflow.name.clone(),
                avatar_url: self.repository.owner.avatar_url.clone(),
            }));
        }

        Ok(None)
    }
}

File: parser/workflowRun.go
package parser

import (
	"fmt"
	"slices"
	"webhook/config"
	"webhook/structs"
)

type workflowRun struct {
	Action      string              `json:"action"`
	Workflow    structs.Workflow    `json:"workflow"`
	WorkflowRun structs.WorkflowRun `json:"workflowRun"`
	Repository  structs.Repository  `json:"repository"`
}

func (e *workflowRun) handle() (*structs.Webhook, error) {
	if e.Action != "completed" || e.WorkflowRun.Conclusion == "" {
		return nil, nil
	}

	if slices.Contains(config.Get().IgnoredWorkflows, e.Workflow.Name) {
		return nil, nil
	}

	emoji := config.Get().SuccessEmoji
	if e.WorkflowRun.Conclusion == "failure" {
		emoji = config.Get().FailureEmoji
	}

	return &structs.Webhook{
		Content: fmt.Sprintf(
			"%s Workflow [%s](<%s>) on [%s](<%s>)/[%[6]s](<%[5]s/tree/%[6]s>)",
			emoji,
			e.WorkflowRun.Conclusion,
			e.WorkflowRun.HtmlUrl,
			e.Repository.Name,
			e.Repository.HtmlUrl,
			e.WorkflowRun.HeadBranch,
		),
		Username:  e.Workflow.Name,
		AvatarUrl: e.Repository.Owner.AvatarURL,
	}, nil
}
